### PR TITLE
[SYCL] Upstream Merge Fixes

### PR DESCRIPTION
**Pull request by agozillon**
Most of the deletions come from some old scheduler files that were not automatically removed in the initial merge. 

Other than that there is the usual culprits for minor fixes, SPIR builtins, reqd-work-group-size, memory_manager tweaks etc. 

The main addition comes from 2 new existing issues tests and I have documented a little bit of my initial investigation inside of: https://github.com/triSYCL/sycl/issues/40

And a new test script for running our tests, it's a stop-gap until someone can spend the time to work out the kinks of integrating our tests into the LLVM lit test environment (Injecting our required environment, adding an option to optionally compile them, possibly annotating some of the tests for proper usage with LLVM lit and adding the correct compiler commands to the head of each file) . But it's mostly there because I got irritated running the tests by hand and it's somewhat useful.

I also altered the sycl-xocc driver based on a bug I found with the new test script.. wasn't playing nice when compiling a file in another directory. Incremental improvements as required.

With this the tests should run but it's currently REQUIRED to have DISABLE_INFER_AS in your environment. e.g. DISABLE_INFER_AS=1

Otherwise your code will work fine for sw_emu but crash in hw_emu in a lot of cases, some generic address space pollution I think. I need to investigate the new address space inference stuff a little bit deeper. This patch is to fix the next branch so that it's use-able to a state that the master branch currently is.

@keryell @Victorlzd I'd appreciate if one of you could run the new test script on 2019.1 hw_emu and see if the tests all play nice. I cannot execute hw_emu tests from 2019.1 on my machine right now, I do most of my hw_emu tests on 2018.3 still.

The hierarchical parallelism stuff will not work right now (as we deemed it unnecessary/low priority at the moment), some SPIRV intrinsic still need SPIR equivalents implemented, mainly barriers. I have opened a new issue on this for new comers to the project as it's a good one for learning about the existing implementation without having to implement any new concepts (hopefully).

### CODE REVIEW

File: sycl/test/xocc_tests/run_tests.sh [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+#   internal_defines
+#
+# These test are the "main" functionality that could break on upstream merge,
+# most of the rest is icing on the cake tests that are related to testing SYCL
+# functionality. They're also important, but larger breaking issues are more
+# likely to show in one of the above tests.
+#
+# Most of the tests will assert and stop the shell running right now. These
+# could perhaps be changed to something like a throw that's caught and prints an
+# error message.. which would perhaps be more user friendly.
```
Comment (keryell) [inline sycl/test/xocc_tests/run_tests.sh]: .. ?

File: sycl/include/CL/sycl/detail/queue_impl.hpp [hunk 2]
```diff
@@ -139,7 +139,15 @@ class queue_impl {
     RT::PiQueue Queue;
     RT::PiContext Context = detail::getSyclObjImpl(m_Context)->getHandleRef();
     RT::PiDevice Device = detail::getSyclObjImpl(m_Device)->getHandleRef();
-#ifdef CL_VERSION_2_0
+  // NOTE: This is the one location CL_VERSION_2_0 is used at the moment
+  // and the CL_VERSION_2_0 seems to come from space, it's not defined by Clang
+  // it defaults to CL_VERSION_2_0 via one of the CL headers even if we undefine
+  // CL_VERSION_2_0 inside of Clangs InitPreprocessor and define CL_VERSION_1_2
+  // instead inside (when it's a xilinx device) this CL_VERSION_2_0 variable
```
Comment (keryell) [inline sycl/include/CL/sycl/detail/queue_impl.hpp]: Xilinx

File: sycl/test/xocc_tests/issue_related/agdce_ice.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    auto Buffer =
+        buffer<point<int>, 1>(Data, range<1>(Size), {property::buffer::use_host_ptr()});
+    Buffer.set_final_data(nullptr);
+    selector_defines::CompiledForDeviceSelector selector;
+    queue Queue {selector};
+    Queue.submit([&](handler &Cgh) {
+      accessor<point<int>, 1, access::mode::write, access::target::global_buffer>
+          Accessor(Buffer, Cgh, range<1>(Size));
+              Cgh.parallel_for<ice_kernel>(range<1>{Size},
+                                         [=](id<1> Index) {
```
Comment (keryell) [inline sycl/test/xocc_tests/issue_related/agdce_ice.cpp]: Indent

File: sycl/test/xocc_tests/issue_related/boost_hana_functor_arg.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    auto a_w = a.get_access<access::mode::write>();
+    for (int i = 0; i < N; ++i) {
+      a_w[i] = i;
+    }
+  }
+
+  auto event = q.submit([&](handler &cgh) {
+      auto a_rw = a.get_access<access::mode::read_write>(cgh);
+
+      cgh.single_task<class array_add>([=]() {
```
Comment (keryell) [inline sycl/test/xocc_tests/issue_related/boost_hana_functor_arg.cpp]: Useless `()`

File: sycl/test/xocc_tests/run_tests.sh [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+# have to visually parse for now. It's not intended to be a replacement for
+# the more complex and useful LLVM tests. It's a stopgap script till time can be
+# put into integrating these tests into the suite and adding the option of
+# running all the other LLVM SYCL tests for FPGA (i.e the ones that Intel and
+# other contributors have made).
+#
+# Example invocation:
+# ./run_tests.sh /my/clang/build/bin optional_output_file.txt
+#
+# Note: You will of course have to have your enviornment setup to compile for
```
Comment (keryell) [inline sycl/test/xocc_tests/run_tests.sh]: environment

File: sycl/test/xocc_tests/run_tests.sh [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    local RUNTIME_ARG_ARR_REF=""
+  fi
+
+  echo "" >>  $TEST_OUTPUT_FILE
+  echo "" >>  $TEST_OUTPUT_FILE
+  echo "Compiling $2/$1.cpp" >>  $TEST_OUTPUT_FILE
+  $CLANG_BIN/clang++ "${USED_DEFAULT_ARGS[@]}" "${COMPILATION_ARG_ARR_REF[@]}" \
+    "$2/$1.cpp" -o "$2/$1.$XCL_EMULATION_MODE" >> $TEST_OUTPUT_FILE 2>&1
+
+  # The default is not to run for HW, to run tests for hw remove the if block,
```
Comment (keryell) [inline sycl/test/xocc_tests/run_tests.sh]: HW or hw? Both are used on this line

File: sycl/test/xocc_tests/run_tests.sh (hunk 1 above)
Comment (agozillon) [reply sycl/test/xocc_tests/run_tests.sh]: Assertions will kill the shell script, as it sends a SIGABRT I think. But it should probably be phrased as "Most of the tests will assert when they are incorrect and fail which will stop the shell script running.."

**Comment by keryell** [comment]
Thanks a lot!

**Pull request merged**